subsys/bluetooth/services/fast_pair/fp_storage/fp_storage.c

```diff
@@ -313,9 +317,54 @@ void fp_storage_ram_clear(void)
 
 	settings_set_err = 0;
 	atomic_set(&settings_loaded, false);
+	reset_in_progress = false;
 }
 
 bool bt_fast_pair_has_account_key(void)
 {
 	return (fp_storage_account_key_count() > 0);
 }
+
+static int fp_storage_reset(void)
+{
+	for (size_t index = 0; index < CONFIG_BT_FAST_PAIR_STORAGE_ACCOUNT_KEY_MAX; index++) {
+		int err;
+		int n;
+		char name[SETTINGS_AK_NAME_MAX_SIZE];
+
+		n = snprintf(name, SETTINGS_AK_NAME_MAX_SIZE, "%s%u", SETTINGS_AK_FULL_PREFIX,
+			     index);
+		__ASSERT_NO_MSG(n < SETTINGS_AK_NAME_MAX_SIZE);
+		if (n < 0) {
+			return n;
+		}
+
+		err = settings_delete(name);
+		if (err) {
+			return err;
+		}
+	}
+
+	memset(account_key_list, 0, sizeof(account_key_list));
+	memset(account_key_loaded_ids, 0, sizeof(account_key_loaded_ids));
+	account_key_next_id = ACCOUNT_KEY_MIN_ID;
+	account_key_count = 0;
+
+	settings_set_err = 0;
+	atomic_set(&settings_loaded, true);
+
+	reset_in_progress = false;
+
+	return 0;
+}
+
+static int reset_in_progress_set(void)
+{
+	reset_in_progress = true;
+
+	return 0;
+}
+
+FP_STORAGE_MANAGER_MODULE_REGISTER(fp_storage_ak, fp_storage_reset, reset_in_progress_set,
```

Comment:
I don't agree. The single macro approach enforces the user who develops the FP storage module to register it correctly for Settings storage and factory reset functionality.

Reply:
Submodule could omit using the fp_storage's macro anyway (and directly use settings API allowing to access non-volatile memory). A submodule may also improperly handle calls from the `fp_storage` main module. Each submodule uses settings API directly to store the data in non-volatile memory, so actually `fp_storage` only notifies on what should by done, but implementation is still up to the submodule. Because of that giving submodule full control over it's non volatile data seems more reasonable. In the future a submodule could also use other unique backend for accessing non-volatile memory without need of API update.

If we would like to change the approach, then maybe we should access settings through the `fp_storage` main module (both store and load). Then main module would have actual control and it could ensure that operations would be handled properly. On the other hand this solution could lead to problems if we would like to introduce a new non volatile memory API usage in the future (like storing keys on some Secure Element). It also introduces new abstraction layer which may not be necessary.

Reply:
Discussed offline. There will be separate macro to register storage module to starage manager and every module can use its own storage backend (in our case it will be Zephyr's settings).